src/coreclr/vm/method.cpp

```diff
@@ -4866,7 +4866,11 @@ bool MethodDesc::DetermineAndSetIsEligibleForTieredCompilation()
         !IsJitOptimizationDisabled() &&
 
         // Policy - Tiered compilation is not disabled by the profiler
-        !CORProfilerDisableTieredCompilation())
+        !CORProfilerDisableTieredCompilation() &&
+
+        // Functional requirement - UnmanagedCallersOnly methods exit in Preemptive GC, which is incompatible with
+        // the lifetime management used for tiered compilation call counting today.
+        !HasUnmanagedCallersOnlyAttribute())
```

Comment:
Can we do this lazily - only once the method is about to be JITed?

Reply:
I'll try to find a place that both makes sure we don't patch the prestub with call counting and set the right optimization tier for the JIT. My guess is we will end up checking in two places (one for each of these cases).

Reply:
The attribute lookup may also be too expensive to do for every method that is called. The check done for precompiled code in `MethodDesc::GetPrecompiledCode` currently uses the caller's GC mode to condition the attribute lookup, that might reduce the cost for jitted cases too, the only caller is `MethodDesc::PrepareILBasedCode` currently, that may be a good place to do the check for both cases.

Reply:
I was afraid of that. I'll try another approach.